### PR TITLE
feat(useIDBKeyval): add `options.serializer`

### LINKED ISSUE #4780: `useIDBKeyval ` | custom serialization as per `useStorage`
### Clear and concise description of the problem

The `useIDBKeyval` integration is great, but I was suprised that there was no way to customise the serialization on write/read.

### Suggested solution

It would be good to bring it up to parity with the regular `useStorage`, as per https://vueuse.org/core/useStorage/#custom-serialization

### Alternative

_No response_

### Additional context

_No response_

### Validations

- [x] Follow our [Code of Conduct](https://github.com/vueuse/vueuse/blob/main/CODE_OF_CONDUCT.md)
- [x] Read the [Contributing Guidelines](https://github.com/vueuse/vueuse/blob/main/CONTRIBUTING.md).
- [x] Read the [docs](https://vueuse.org/guide).
- [x] Check that there isn't already an issue that request the same feature to avoid creating a duplicate.

### PR DESCRIPTION
**Pull request by matthewjumpsoffbuildings**
### Description

Attempt to add `serializer` option to `useIDBKeyval` to bring it more in line with `useStorage` as per https://vueuse.org/core/useStorage/#custom-serialization

Closes #4780

### CODE REVIEW

**Comment by ilyaliao** [comment]
Thank you for your PR. Can you add some tests for this feature to prevent future regressions?

File: packages/integrations/useIDBKeyval/index.ts [hunk 1]
```diff
@@ -64,16 +74,24 @@ export function useIDBKeyval<T>(
...
+        }
       }
       else {
-        data.value = rawValue
+        if (serializer) {
+          data.value = serializer.read(rawValue)
+        }
+        else {
+          data.value = rawValue
+        }
```
Comment (OrbisK) [inline packages/integrations/useIDBKeyval/index.ts]: maybe we should do this via default serializer. Like:

```suggestion
        data.value = serializer.read(rawValue)
```

and

```ts
  const {
    flush = 'pre',
    deep = true,
    shallow = false,
    onError = (e) => {
      console.error(e)
    },
    writeDefaults = true,
    serializer: {
      read: (val)=>val,
      write: (val)=>val, 
    }
  } = options
```

**Review by OrbisK — CHANGES_REQUESTED** [review]

**Comment by acrobid** [comment]
I've encountered this breaking change in production and created minimal reproductions that demonstrate the issue:

Before v10.9.0 (working): https://stackblitz.com/edit/vitejs-vite-cdqkbqje?file=package.json
Current version (broken): https://stackblitz.com/edit/vitejs-vite-hzsdf8sq?file=package.json

This change has been blocking our upgrade from @vueuse/integrations v10.7.2. Since it was released as a minor version update, we didn't anticipate breaking changes.

I came here to propose a serialization option and am glad to see it's already been suggested! The approach in this PR looks great and would solve our use case perfectly.

Is there anything I can do to help move this forward? I'd be happy to:

Test the PR in our production environment
Add test cases for edge cases we've encountered
Help with documentation

Thank you for working on this!

File: packages/integrations/useIDBKeyval/index.ts [hunk 2]
```diff
@@ -4,7 +4,12 @@ import { watchPausable } from '@vueuse/core'
 import { del, get, set, update } from 'idb-keyval'
 import { ref as deepRef, shallowRef, toRaw, toValue } from 'vue'
 
-export interface UseIDBOptions extends ConfigurableFlush {
+interface Serializer<T> {
+  read: (raw: any) => T
+  write: (value: T) => any
```
Comment (OrbisK) [inline packages/integrations/useIDBKeyval/index.ts]: maybe we need one more generic for the Raw Values

File: packages/integrations/useIDBKeyval/index.ts [hunk 3]
```diff
@@ -90,8 +108,11 @@ export function useIDBKeyval<T>(
         await del(key)
       }
       else {
-        // IndexedDB does not support saving proxies, convert from proxy before saving
-        await update(key, () => toRaw(data.value))
+        let rawValue = toRaw(data.value)
+        if (serializer) {
+          rawValue = serializer.write(rawValue as any)
```
Comment (OrbisK) [inline packages/integrations/useIDBKeyval/index.ts]: we should replace the cast by the generic

**Review by OrbisK — CHANGES_REQUESTED** [review]

**Comment by OrbisK** [comment]
@matthewjumpsoffbuildings lets get this over the line! Would you be willing to add the suggestions?

**Comment by OrbisK** [comment]
@acrobid You can use `pnpm patch` command to test whether this approach works for you. I think we might merge this soon. But I think we will not backport this to older major versions.

**Comment by matthewjumpsoffbuildings** [comment]
@OrbisK @acrobid I pushed some commits attempting to apply the suggestions, default serializer, better typing etc

I don't have time to write tests right now, if someone else can handle that that would be good

File: packages/integrations/useIDBKeyval/index.ts (hunk 2 above)
Comment (43081j) [reply packages/integrations/useIDBKeyval/index.ts]: technically they can be anything, mutated outside of this code. so i don't think we can safely type them

they really are `unknown` since third party libs can save whatever in there, so we should push serializer implementors to properly deal with it
